tutorials/v7/ntuple/ntpl014_framework.C

```diff
@@ -0,0 +1,409 @@
+};
+
+// An Outputter provides the interface to fill DataProducts into an RNTuple.
+class Outputter {
+public:
+   virtual ~Outputter() = default;
+
+   virtual void InitSlot(unsigned slot) = 0;
+   virtual void Fill(unsigned slot, const std::vector<DataProduct> &products) = 0;
+};
+
+// A ParallelOutputter uses an RNTupleParallelWriter to append an RNTuple to a TFile.
+class ParallelOutputter final : public Outputter {
+   FileService &fFileService;
+   std::unique_ptr<RNTupleParallelWriter> fParallelWriter;
+   std::vector<REntry::RFieldToken> fTokens;
+
+   struct SlotData {
+      std::shared_ptr<RNTupleFillContext> fillContext;
+      std::unique_ptr<REntry> entry;
+   };
+   std::vector<SlotData> fSlots;
+
+public:
+   ParallelOutputter(ModelTokensPair modelTokens, FileService &fileService, std::string_view ntupleName,
+                     const RNTupleWriteOptions &options)
+      : fFileService(fileService), fTokens(std::move(modelTokens.second))
+   {
+      auto &model = modelTokens.first;
+
+      std::lock_guard g(fileService.GetMutex());
+      fParallelWriter = RNTupleParallelWriter::Append(std::move(model), ntupleName, fFileService.GetFile(), options);
+   }
+
+   void InitSlot(unsigned slot) final
+   {
+      if (slot >= fSlots.size()) {
+         fSlots.resize(slot + 1);
+      }
+      // Create an RNTupleFillContext and REntry that are used for all fills from this slot. We recommend creating a
+      // bare entry if binding all fields.
+      fSlots[slot].fillContext = fParallelWriter->CreateFillContext();
+      fSlots[slot].entry = fSlots[slot].fillContext->GetModel().CreateBareEntry();
+   }
+
+   void Fill(unsigned slot, const std::vector<DataProduct> &products) final
+   {
+      assert(slot < fSlots.size());
+      auto &fillContext = *fSlots[slot].fillContext;
+      auto &entry = *fSlots[slot].entry;
+
+      // Use the field tokens to bind the products' raw pointers.
+      for (auto &&product : products) {
+         entry.BindRawPtr(fTokens[product.index], product.address);
+      }
+
+      // Fill the entry without triggering an implicit flush.
+      RNTupleFillStatus status;
+      fillContext.FillNoFlush(entry, status);
+      if (status.ShouldFlushCluster()) {
+         // If we are asked to flush, first try to do as much work as possible outside of the critical section:
+         // FlushColumns() will flush column data and trigger compression, but not actually write to storage.
+         // (A framework may of course also decide to flush more often.)
+         fillContext.FlushColumns();
+
+         {
+            // FlushCluster() will flush data to the underlying TFile, so it requires synchronization.
+            std::lock_guard g(fFileService.GetMutex());
+            fillContext.FlushCluster();
+         }
+      }
+   }
+};
+
+// A SerializingOutputter uses a sequential RNTupleWriter to append an RNTuple to a TFile and a std::mutex to
+// synchronize multiple threads.
```

Comment:
Maybe we could add a note about IMT (or enable it), to show that page compression can still be in parallel even with the sequential mode.

Reply:
I thought about this and unfortunately I think this is not possible / not efficient with the implementation in this tutorial: A thread blocking to acquire a `std::mutex` cannot "help" the other thread that is currently in the critical section by executing its tasks. I added a note, also on the direct usage of `std::thread`.